users/postgres/groups.go

```diff
@@ -0,0 +1,497 @@
+			return users.GroupPage{}, err
+		}
+
+		items = append(items, gr)
+	}
+
+	cq := fmt.Sprintf(`SELECT COUNT(*) FROM groups WHERE owner_id = :owner_id
+	AND parent_id in (select id from groups where id = :id)  %s;`, mq)
+
+	total, err := total(ctx, gr.db, cq, dbPage)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := users.GroupPage{
+		Groups: items,
+		PageMetadata: users.PageMetadata{
+			Total:  total,
+			Offset: offset,
+			Limit:  limit,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) RetrieveAllForUser(ctx context.Context, userID string, offset, limit uint64, gm users.Metadata) (users.GroupPage, error) {
+	m, mq, err := getMetadataQuery(gm)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+
+	q := fmt.Sprintf(`SELECT g.id, g.owner_id, g.parent_id, g.name, g.description, g.metadata 
+					  FROM group_relations gr, groups g
+					  WHERE gr.group_id = g.id and gr.user_id = :userID
+		  			  %s ORDER BY id LIMIT :limit OFFSET :offset;`, mq)
+
+	params := map[string]interface{}{
+		"userID":   userID,
+		"limit":    limit,
+		"offset":   offset,
+		"metadata": m,
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, params)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	var items []users.Group
+	for rows.Next() {
+		dbgr := dbGroup{}
+		if err := rows.StructScan(&dbgr); err != nil {
+			return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+		}
+
+		gr := toGroup(dbgr)
+		if err != nil {
+			return users.GroupPage{}, err
+		}
+
+		items = append(items, gr)
+	}
+	cq := fmt.Sprintf(`SELECT COUNT(*) 
+					   FROM group_relations gr, groups g
+					   WHERE gr.group_id = g.id and gr.user_id = :userID %s;`, mq)
+
+	total, err := total(ctx, gr.db, cq, params)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := users.GroupPage{
+		Groups: items,
+		PageMetadata: users.PageMetadata{
+			Total:  total,
+			Offset: offset,
+			Limit:  limit,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) AssignUser(ctx context.Context, userID, groupID string) error {
+	tx, err := gr.db.BeginTxx(context.Background(), nil)
+	if err != nil {
+		return errors.Wrap(errDeleteGroupDB, err)
+	}
+	defer func() {
+		if err != nil {
+			if txErr := tx.Rollback(); txErr != nil {
+				err = errors.Wrap(err, errors.Wrap(errTransRollback, txErr))
+			}
+		}
+
+		if err = tx.Commit(); err != nil {
+			err = errors.Wrap(errDeleteGroupDB, err)
+		}
+	}()
+
+	q := `SELECT COUNT(*) FROM group_relations WHERE group_id = :group_id AND user_id = :user_id ;`
+	dbr, err := toDBGroupRelation(userID, groupID)
+	if err != nil {
+		return errors.Wrap(users.ErrAssignUserToGroup, err)
+	}
+	tot, err := total(ctx, gr.db, q, dbr)
+	if err != nil {
+		return errors.Wrap(users.ErrAssignUserToGroup, err)
+	}
+	if tot > 0 {
+		return errors.Wrap(users.ErrUserAlreadyAssigned, err)
+	}
```

Comment:
The same goes here. We can simply try to insert, and examine an error to decide what went wrong.